### DISCUSSION TITLE
Maximum diversification for corporate bonds

### DISCUSSION THREAD

**Discussion by nozomi00** [skfolio/skfolio#73] [Q&A] [answered]
Hi,

I am working on my thesis that researches on the potential of a buy-and-maintain corporate bond portfolio. One of the measures I use is the diversification ratio. As skfolio already provides the function 'MaximumDiversification' which maximizes the DR for a given set of assets and can return the optimal weights for these assets. One of my questions is that: given the investable universe (800 bonds you can choose from), is it able to use this one to reach an optimal portfolio of size 5 (only 5 assets) that has the highest DR among all the possible combinations of 5 bonds. 

I noticed there are ways to add constraints in the maximization problem, but it seems all these constraints are on weights ( so I can constraint a particular asset's weight or a combination of certain assets' weight). It'll be nice if someone could confirm whether the in-built optimizer indeed cannot solve my problem (selecting 5 out of many bonds to form the optimal portfolio), or if someone happens to know some good methodologies to address my problem. 

PS: At the moment, I am constructing these optimal portfolios in a heuristic way, such that the first one is picked randomly and the rest is selected when it gives the maximum marginal diversification, until the portfolio size reaches 5. (of course in this way we can never ensure to get a global optimum)

Best,
Xi

**Comment by HugoDelatte** [comment, marked as answer]
Hello,

You need cardinality constraint which is a Mixed Integer Programming problem (MIP).
To solve it, you can use the parameter `add_constraints` that lets you add custom constraints. You will also need to use a solver that supports MIP such as `SCIP` by installing it (pip install cvxpy[SCIP]).

```python
from skfolio.datasets import load_sp500_dataset
from skfolio.preprocessing import prices_to_returns
from skfolio.optimization import MeanRisk
import cvxpy as cp

prices = load_sp500_dataset()
X = prices_to_returns(prices)

n_observations, n_assets = X.shape
card = 5


def card_constraints(w):
    card_var = cp.Variable(20, boolean=True)
    constraints = [
        w - card_var <= 0,
        cp.sum(card_var) <= card,
    ]
    return constraints


model = MeanRisk(add_constraints=card_constraints, solver="SCIP")
model.fit(X)
assert sum(model.weights_ != 0) == card

```

Note that open source solvers may fail more often to solve MIP problems as opposed to  `Mosek` because of the difficulty linked to MIP.

If your problem fails, you have multiple workarounds:
* use an heuristic as you mentioned (solve for all combination of 5 assets)
*  use weight regularization (check the parameters `l2_coef` and `l1_coef`). On long only portfolio, L2 regularization will send some weights near 0. You can then remove the assets with weights below a given threshold and eventually re-run an optimization without regularization.

**Reply by nozomi00** [reply]
Hi Hugo,

A big thank you for the weight regularization! Evaluated out of sample, these portfolios perform better than the heuristically optimized ones.

**Reply by HugoDelatte** [reply]
I'm glad it worked.
For weights regularization, a good approach is to use `RandomizedSearchCV` to find the regularization params that maximizes (or minimizes) a given score evaluated out-of-sample. You can find an example [here](https://skfolio.org/auto_examples/1_mean_risk/plot_8_regularization.html#hyper-parameter-tuning)

**Comment by HugoDelatte** [comment]
@nozomi00 v0.6.0 now contains cardinality , group cardinality and threshold constraints and is available for ratio too (ex: Sharpe Ratio). Special calibration has been implemented to improve convergence.
You can use these two examples:

https://skfolio.org/auto_examples/1_mean_risk/plot_15_mip_cardinality_constraints.html#sphx-glr-auto-examples-1-mean-risk-plot-15-mip-cardinality-constraints-py

https://skfolio.org/auto_examples/1_mean_risk/plot_16_mip_threshold_constraints.html#sphx-glr-auto-examples-1-mean-risk-plot-16-mip-threshold-constraints-py
